### PR TITLE
Integrate OSSAR (Open Source Static Analysis Runner)

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Use the OSSAR action to verify binaries are built with correct compiler / linker options.

## Testing

CI/C

## Documentation

No.

### CODE REVIEW

File: .github/workflows/ossar-scan.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  contents: read  # Required by actions/checkout to fetch code.
+  security-events: write # Required by codeql task
+
+jobs:
+  scan:
+    timeout-minutes: 30
+
+    strategy:
+      matrix:
+        # For now only run on release as debug builds of the MSVC are built with out /Qspectre
```
Comment (dthaler) [inline .github/workflows/ossar-scan.yml:31]:
```suggestion
        # For now only run on release as debug builds of the MSVC are built without /Qspectre
```

File: Directory.Build.props [hunk 2]
```diff
@@ -22,7 +25,12 @@
       <SupportJustMyCode>false</SupportJustMyCode>
       <DebugInformationFormat>ProgramDatabase</DebugInformationFormat>
       <IntrinsicFunctions>true</IntrinsicFunctions>
+      <ControlFlowGuard>Guard</ControlFlowGuard>
+      <AdditionalOptions>/ZH:SHA_256  %(AdditionalOptions)</AdditionalOptions>
```
Comment (dthaler) [inline Directory.Build.props:29]:
```suggestion
      <AdditionalOptions>/ZH:SHA_256 %(AdditionalOptions)</AdditionalOptions>
```

**Review by dthaler — APPROVED** [review]

File: .github/workflows/ossar-scan.yml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+jobs:
+  scan:
+    timeout-minutes: 30
+
+    strategy:
+      matrix:
+        # For now only run on release as debug builds of the MSVC are built without /Qspectre
+        configurations: [Release]
+    runs-on: windows-latest
```
Comment (saxena-anurag) [inline .github/workflows/ossar-scan.yml:33]: Should this be `windows-2019`? We are using `windows-2019` at other places.

Comment (Alan-Jowett) [reply .github/workflows/ossar-scan.yml:33]: The job doesn't run on Windows-2019 due to .NET issue. Only works on Windows-Latest:
https://github.com/github/ossar-action

Comment (saxena-anurag) [reply .github/workflows/ossar-scan.yml:33]: Maybe we can add that as a comment in yml.

**Review by saxena-anurag — DISMISSED** [review]

**Review by saxena-anurag — APPROVED** [review]

File: .github/workflows/ossar-scan.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+jobs:
+  scan:
+    timeout-minutes: 30
+
+    strategy:
+      matrix:
+        # For now only run on release as debug builds of the MSVC are built without /Qspectre
+        configurations: [Release]
+    # github/ossar-action doesn't run on window-2019, requires windows-latest.
```
Comment (dthaler) [inline .github/workflows/ossar-scan.yml:33]:
```suggestion
    # github/ossar-action doesn't run on windows-2019, requires windows-latest.
```

File: .github/workflows/ossar-scan.yml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      run: |
+        pushd ${{github.workspace}}/${{env.BUILD_PLATFORM}}/${{env.BUILD_CONFIGURATION}}
+        Remove-Item @("clang_rt.*", "concrt*", "msvc*", "ucrt*", "vc*") -ErrorAction SilentlyContinue
+
+    - name: Run OSSAR (Open Source Static Analysis Runner)
+      uses: github/ossar-action@c757d32d66bea728bc64e67e7d6de9696f7f37d3
+      id: ossar
+
+    - name: Upload results to Security tab
+      uses: github/codeql-action/upload-sarif@v1
```
Comment (dthaler) [inline .github/workflows/ossar-scan.yml:75]: I suspect you'll get a "please pin to a hash instead" issue about this line

**Review by dthaler — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Comment by Alan-Jowett** [comment]
> Is there a badge we can add to the top of the README.md file? https://github.com/github/ossar-action shows 2 such badges on that page, so I expect we should be able to.

I don't see how to do that currently. It appears as if badges can only be created for the top-level workflow (CI/CD) and not for jobs within a workflow.

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**